### ISSUE TITLE
add usearch's fastq_mergepairs function and options

### ISSUE THREAD

**Issue by dereneaton** [torognes/vsearch#44]
I had been using a function in usearch to detect overlap of paired-end reads. This used the following options which would be nice to have available in vsearch. 

```
usearch_v.7.0.1090_i86linux32 -fastq_mergepairs read1.fq 
                              -reverse read2.fq
                              -fastq_maxdiffs 6
                              -fastq_truncqual 2 
                              -fastq_minlen 36 
                              -fastq_minmergelen 50 
                              -fastqout merge.fq 
                              -fastq_allowmergestagger
```

**Comment by frederic-mahe** [comment]
That [Biostar topic](https://www.biostars.org/p/144950/) lists C/C++ open source software parsing or manipulating FASTQ files. Maybe there are some good ideas to grab there.

**Comment by colinbrislawn** [comment]
Hello @frederic-mahe!

Oh there are lots of good processing tools. I get good use out of [ea-utils](https://code.google.com/p/ea-utils/) and there is a compairson of tools [here](https://www.biostars.org/p/85929/#86082).

I believe USEARCH is so successful because it does _literally everying_ you could want to do with NGS data. This is why it's the only closed-source programs in my pipeline. VSEARCH now covers most parts of an amplicon study and those preprocessing steps are the last pieces missing. 

I know this functionality is redundant, and working on [local alignments](http://www.drive5.com/usearch/manual/cmd_usearch_local.html) would probably be more fun. I still value one stop shopping, and I think having @xflouris implement PEAR as `-fastq_mergepairs` would be really helpful.

I love how comprehensive VSEARCH is already and look forward to future development. 
Colin

**Comment by xflouris** [comment]
Hi Colin,
I'm currently re-structuring the PEAR source code to make it manageable, as it was developed more as a toy implementation by several people with different coding styles. We also have a new, non-heuristic and exact way of computing the posterior quality scores for the merged part of the reads, which will be available in the next version of PEAR (0.9.9). 

Once this is done, it will be very easy to incorporate it in VSEARCH in the form of a -fastq_mergepairs, as in USEARCH.

@torognes, @frederic-mahe  I can incorporate the algorithm during summer, if you agree.

Tomas

**Comment by torognes** [comment]
@xflouris, that sounds brilliant!

**Comment by colinbrislawn** [comment]
@xflouris Wow that's even better than I expected! I especially appreciate this team's dedication to using exact algorithms. I'm looking forward to it!

**Comment by xflouris** [comment]
Hi @torognes I've scheduled my time to work on fastq_mergepairs from 10.8 to 21.8.
Cheers, tomas

**Comment by torognes** [comment]
A new FASTQ file parser has been added and the fastq_chars, fastq_stats and fastq_filter commands with all options have been added in version 1.2.16 in the summer branch. Unable to find out what the number in the "Tails" column in fastq_chars output is supposed to mean.

**Comment by chrismclimans** [comment]
@torognes that is great. I have, in the past 3 months, started to implement VSEARCH into my lab's NGS pipeline and we are throughly enjoying using it. Like Colin said we still rely on USEARCH for -fastq_stats -fastq_mergepairs and -fastq_filter. I'm excited that some of these have been added to the new version. When do you foresee releasing the newest (1.2.16) version?

**Comment by colinbrislawn** [comment]
@chrismclimans You can download, build, and use the newest right now, if you want to. 
```
git clone https://github.com/torognes/vsearch.git
cd vsearch
git checkout summer
./autogen.sh
./configure
make
```

The new `vsearch` executable will be in the folder `bin`.

@xflouris Any updated on `--fastq_mergepairs`? I would love to see that in vsearch!

Keep up the great work.
Colin

**Comment by chrismclimans** [comment]
@colinbrislawn Thanks so much Colin, hope you are doing well! (-from the Lamendella lab).

**Comment by torognes** [comment]
@chrismclimans Nice to hear that you enjoy vsearch. We need to do some more testing, documentation and repo preparation before I think we can make a new release. I think it will be available later in September.

**Comment by torognes** [comment]
fastx_revcomp has also been added.

**Comment by torognes** [comment]
I've started work on fastq_mergepairs.

**Comment by colinbrislawn** [comment]
Excellent! I look forward to this functionality.

What method will you use to calculate Q scores in the resulting area of overlap?
@xflouris mentioned using an improved metric. I know `bwa pemerge` and `PEAR` use different metrics if I remember correctly. I have no idea what `seqprep` uses...

**Comment by colinbrislawn** [comment]
The method USEARCH uses for calculating post-pairing Q scores, is described this paper:
Section 2.4 Posterior error probability for a merged base call http://bioinformatics.oxfordjournals.org/content/early/2015/07/29/bioinformatics.btv401.full.pdf

**Comment by torognes** [comment]
Thanks, @colinbrislawn, I'll probably use the equations in Edgar & Flyvbjerg's paper for computing the quality scores of merged bases.

**Comment by torognes** [comment]
The `--fastq_mergepairs` command is implemented in version 1.9.0 that was just released. All options in usearch are supported. Statistical testing of significance is performed in a way similar to [PEAR](https://github.com/xflouris/PEAR), but it does not use empirically determined base frequencies and just assumes 25% of each base. The quality of merged bases is computed as described by [Edgar and Flyvbjerg (Bioinformatics, 2015)](10.1093/bioinformatics/btv401). It is not parallelised yet. It should be considered experimental for now.

**Comment by torognes** [comment]
The following commands to process FASTQ files are now available in vsearch:
* fastq_chars
* fastq_convert
* fastq_filter
* fastq_mergepairs
* fastq_stats
* fastx_mask
* fastx_revcomp
* fastx_subsample

They behave is with usearch. The new fastq_convert command can convert FASTQ files between different quality score encodings (phred+33 and phred+64).

Please see the [manual](https://github.com/torognes/vsearch/releases/download/v1.9.0/vsearch_manual.pdf) for details.

**Comment by colinbrislawn** [comment]
:clap: Thank you! I look forward to using these!

**Comment by torognes** [comment]
The fastq_mergepairs command is not yet parallelised. I plan to do this later.

**Comment by torognes** [comment]
This issue should be solved in version 1.10.0 just released. The fastq_mergepairs command is now parallelized and very fast. It's been better tested as well.

**Comment by zkanfra** [comment]
hello all, 
I am new here and would like to know how to merge all my Forward and reverse reads all at a go using usearch. 
I used the script   usearch -fastq_mergepairs *R1*.fastq -fastqout merged.fq but doesnt work on windows.. 
Please can someone advice me on what to do

**Comment by torognes** [comment]
Hi @zkanfra!

This the repository for vsearch, not for usearch. You can find usearch here: http://www.drive5.com/usearch/

For support questions I would suggest that you use the VSEARCH Forum, here: https://groups.google.com/forum/#!forum/vsearch-forum

If you want to merge reads with vsearch you should use the following command:

`vsearch --fastq_mergepairs R1.fastq --reverse R2.fastq --fastqout merged.fq`

You need to specify the FASTQ files with the R2 reads with the `--reverse` option.

I hope that works for you.

**Comment by MiqG** [comment]
Dear all, thank you for developing such a great and complete software!!

I think I have found a small bug in `--fastq_mergepairs`  from `vsearch v2.14.2_linux_x86_64`, as the option `--fastq_qminout` is not available in the command although it appears in the documentation. See error:
```
  vsearch --fastq_mergepairs fw.fq \
                --reverse rv.fq \
                --fastqout merged.fq \
                --threads 4 \
                --log vsearch.log \
                --fastq_allowmergestagger \
                --fastq_maxmergelen 100 \
                --fastq_minmergelen 100 \
                --fastq_qminout 25 \
                > vsearch.o
```
```
Fatal error: Invalid options to command fastq_mergepairs
Invalid option(s): --fastq_qminout
The valid options for the fastq_mergepairs command are: --bzip2_decompress --eeout --eetabbedout --fasta_width --fastaout --fastaout_notmerged_fwd --fastaout_notmerged_rev --fastq_allowmergestagger --fastq_ascii --fastq_eeout --fastq_maxdiffpct --fastq_maxdiffs --fastq_maxee --fastq_maxlen --fastq_maxmergelen --fastq_maxns --fastq_minlen --fastq_minmergelen --fastq_minovlen --fastq_nostagger --fastq_qmax --fastq_qmaxout --fastq_qmin --fastq_truncqual --fastqout --fastqout_notmerged_fwd --fastqout_notmerged_rev --gzip_decompress --label_suffix --log --no_progress --quiet --relabel --relabel_keep --relabel_md5 --relabel_self --relabel_sha1 --reverse --sizein --sizeout --threads --xee --xsize
```
See documentation (with `vsearch -h`):
```
Paired-end reads merging
  --fastq_mergepairs FILENAME merge paired-end reads into one sequence
 Data
  --reverse FILENAME          specify FASTQ file with reverse reads
 Parameters
  --fastq_allowmergestagger   allow merging of staggered reads
  --fastq_ascii INT           FASTQ input quality score ASCII base char (33)
  --fastq_maxdiffpct REAL     maximum percentage diff. bases in overlap (100.0)
  --fastq_maxdiffs INT        maximum number of different bases in overlap (10)
  --fastq_maxee REAL          maximum expected error value for merged sequence
  --fastq_maxmergelen         maximum length of entire merged sequence
  --fastq_maxns INT           maximum number of N's
  --fastq_minlen INT          minimum input read length after truncation (1)
  --fastq_minmergelen         minimum length of entire merged sequence
  --fastq_minovlen            minimum length of overlap between reads (10)
  --fastq_nostagger           disallow merging of staggered reads (default)
  --fastq_qmax INT            maximum base quality value for FASTQ input (41)
  --fastq_qmaxout INT         maximum base quality value for FASTQ output (41)
  --fastq_qmin INT            minimum base quality value for FASTQ input (0)
  --fastq_qminout INT         minimum base quality value for FASTQ output (0)
  --fastq_truncqual INT       base quality value for truncation
 Output
  --eetabbedout FILENAME      output error statistics to specified file
  --fastaout FILENAME         FASTA output filename for merged sequences
  --fastaout_notmerged_fwd FN FASTA filename for non-merged forward sequences
  --fastaout_notmerged_rev FN FASTA filename for non-merged reverse sequences
  --fastq_eeout               include expected errors (ee) in FASTQ output
  --fastqout FILENAME         FASTQ output filename for merged sequences
  --fastqout_notmerged_fwd FN FASTQ filename for non-merged forward sequences
  --fastqout_notmerged_rev FN FASTQ filename for non-merged reverse sequences
  --label_suffix              suffix to append to label of merged sequences
  --xee                       remove expected errors (ee) info from output
```

Best

**Comment by torognes** [comment]
Thanks for reporting this. It seems like the `--fastq_qminout` is missing, yes. I'll add it.

**Comment by thomasvangurp** [comment]
@torognes thanks for this amazing tool! a small bug to add:
when using `--relabel_keep --label_suffix "\tRG:Z:my_rg_id` i get the read group id successfully in the file specified by `--fastqout` but not for `--fastqout_notmerged_fwd` and `--fastqout_notmerged_rev`

**Comment by torognes** [comment]
@thomasvangurp Thanks for reporting this issue. I'll fix it in the next release. I've been a bit busy with other things lately, but it seems like I'll get more time for vsearch from now on.

**Comment by torognes** [comment]
The missing `fastq_qminout` option has been added in commit e34d5894f28ee5e8479c69556a90c912ea6a3ff1.

**Comment by torognes** [comment]
The problems with `--label_suffix` for `--fastq_mergepairs` should be fixed in commit c55fde0736bd20d662362ef7f1482bab6f9934f2.
